### PR TITLE
Replace ICodeLoader with ICodeDetailsLoader

### PR DESCRIPTION
**Pull request by sonalivdeshpande**
In this PR, the usage of ICodeLoader is replaced with ICodeDetailsLoader in the entire repo. The old interface is not yet removed. 

Closes: https://github.com/microsoft/FluidFramework/issues/9033

### CODE REVIEW

File: BREAKING.md [hunk 1]
```diff
@@ -67,6 +72,27 @@ The `packageInfoSource` argument in `getAbsoluteUrl()` on `@fluidframework/odsp-
 + ): Promise<string>;
 ```
 
+### Replace ICodeLoader with ICodeDetailsLoader interface
+The interface `ICodeLoader` was deprecated a while ago in previous releases. The alternative for `ICodeLoader` interface is the `ICodeDetailsLoader` interface which can be imported from `@fluidframework/container-definitions`. `ICodeLoader` interface will be removed in the next release.
+
+`ILoaderService` and `ILoaderProps` are the two interface are are type to ICodeDetailsLoader only. Please make sure to change their implementation.
```
Comment (ChumpChief) [inline BREAKING.md:78]:
```suggestion
In particular, note the `ILoaderService` and `ILoaderProps` interfaces used with the `Loader` class now only support `ICodeDetailsLoader`.  If you were using an `ICodeLoader` with these previously, you'll need to update to an `ICodeDetailsLoader`.
```

File: packages/loader/container-loader/src/containerContext.ts [hunk 2]
```diff
@@ -329,7 +329,7 @@ export class ContainerContext implements IContainerContext {
                 details: details ?? codeDetails,
             };
         } else {
-            return { module: loadCodeResult, details: codeDetails };
+            return loadCodeResult;
```
Comment (ChumpChief) [inline packages/loader/container-loader/src/containerContext.ts:332]: Currently, this path handles the `IFluidModule` case (forcibly appending a details to it), while the above path handles the `IFluidModuleWithDetails`.  With this change, we don't expect this compat path to execute anymore (`ICodeDetailsLoader` should always result in an `IFluidModuleWithDetails`).

Although your change will result in compile-time breaks that will push people to update their code loaders, I think it might be possible that folks dynamically load their code loaders.  If that can happen, then retaining this runtime handling is probably good, plus instrument to see when that usage goes away.  Probably something like:

```suggestion
            // If "module" is not in the result, we are using a legacy ICodeLoader.  Fix the result up with details.
            // Once usage drops to 0 we can remove this compat path.
            this.taggedLogger.sendTelemetryEvent({ eventName: "LegacyCodeLoader" });
            return { module: loadCodeResult, details: codeDetails };
```

File: packages/loader/container-loader/src/test/containerContext.spec.ts [hunk 3]
```diff
@@ -114,19 +121,27 @@ describe("ContainerContext Tests", () => {
         mockCodeLoader.verify();
     });
 
+    // open an issue to add string typing
```
Comment (ChumpChief) [inline packages/loader/container-loader/src/test/containerContext.spec.ts:124]: I think this was "strict" typing from our conversation last week, please do open an issue.

```suggestion
```

File: packages/loader/container-loader/src/test/containerContext.spec.ts [hunk 4]
```diff
@@ -153,20 +168,28 @@ describe("ContainerContext Tests", () => {
...
```
Comment (ChumpChief) [inline packages/loader/container-loader/src/test/containerContext.spec.ts:171]:
```suggestion
```

File: packages/loader/web-code-loader/src/webLoader.ts [hunk 5]
```diff
@@ -60,10 +61,13 @@ export class WebCodeLoader implements ICodeLoader {
...
 
         const fluidModuleP = this.loadModuleFromResolvedCodeDetails(resolved);
+
         if (resolved.resolvedPackageCacheId !== undefined) {
             this.loadedModules.set(resolved.resolvedPackageCacheId, fluidModuleP);
         }
-        return fluidModuleP;
+        const fluidModule = await fluidModuleP;
+
+        return fluidModule;
```
Comment (ChumpChief) [inline packages/loader/web-code-loader/src/webLoader.ts:70]: Returning the promise should be fine here, since we don't need to modify the module before return.

```suggestion
        return fluidModuleP;
```

File: packages/tools/webpack-fluid-loader/src/loader.ts [hunk 6]
```diff
@@ -92,7 +93,8 @@ export type RouteOptions =
     | ITinyliciousRouteOptions
     | IOdspRouteOptions;
 
-function wrapWithRuntimeFactoryIfNeeded(packageJson: IFluidPackage, fluidModule: IFluidModule): IFluidModule {
+async function wrapWithRuntimeFactoryIfNeeded(packageJson: IFluidPackage, fluidModule: IFluidModule):
```
Comment (ChumpChief) [inline packages/tools/webpack-fluid-loader/src/loader.ts:96]: This seems like it should stay synchronous?

File: examples/hosts/iframe-host/src/inner.ts [hunk 7]
```diff
@@ -51,8 +51,14 @@ async function loadContainer(
     const documentServiceFactory = await InnerDocumentServiceFactory.create(innerPort);
     const urlResolver = await InnerUrlResolver.create(innerPort);
 
-    const module = { fluidExport: TodoContainer };
-    const codeLoader = { load: async () => module };
+    const load = async (): Promise<IFluidModuleWithDetails> => {
+        return {
+            module: { fluidExport: TodoContainer },
+            details: { package: "no-dynamic-package", config: {} },
```
Comment (ssimic2) [inline examples/hosts/iframe-host/src/inner.ts:57]: @ChumpChief seems like we are using fake proposals often throughout the codebase. Should this eventually go away? It doesn't need to be this PR.

File: packages/loader/web-code-loader/src/webLoader.ts [hunk 8]
```diff
@@ -79,7 +83,7 @@ export class WebCodeLoader implements ICodeLoader {
 
         let fluidModule: IFluidModule | undefined;
         for (const script of await loadedScriptsP) {
-            const maybeFluidModule = script.entryPoint as IFluidModule;
+            const maybeFluidModule = script.entryPoint as any;
```
Comment (ssimic2) [inline packages/loader/web-code-loader/src/webLoader.ts:86]: Why case to `any`? What do you expect here?

File: packages/tools/webpack-fluid-loader/src/loader.ts (hunk 6 above)
Comment (ssimic2) [reply packages/tools/webpack-fluid-loader/src/loader.ts:96]: I do not see added async call that would need to make this call return a promise.

File: packages/tools/replay-tool/src/replayLoaderObject.ts [hunk 9]
```diff
@@ -47,8 +48,11 @@ export class ReplayCodeLoader implements ICodeLoader, IFluidCodeDetailsComparer
         return this;
     }
 
-    public async load(source: IFluidCodeDetails): Promise<IFluidModule> {
-        return Promise.resolve(this.fluidModule);
+    public async load(source: IFluidCodeDetails): Promise<IFluidModuleWithDetails> {
+        return {
+            module: this.fluidModule,
+            details: source,
```
Comment (ssimic2) [inline packages/tools/replay-tool/src/replayLoaderObject.ts:54]: It seems `source` can be a string or an object based on logic above. Probably need some handling before being added to details

File: packages/loader/container-loader/src/test/containerContext.spec.ts [hunk 10]
```diff
@@ -153,20 +168,28 @@ describe("ContainerContext Tests", () => {
...
+            load,
             get IFluidCodeDetailsComparer() {
                 return this;
             },
             satisfies: async (
                 candidate: IFluidCodeDetails,
                 constraint: IFluidCodeDetails,
-            ) => {},
+            ) => { return true; },
+            compare: async () => { return 0; },
```
Comment (ssimic2) [inline packages/loader/container-loader/src/test/containerContext.spec.ts:192]: What is the purpose of this change?

File: packages/tools/webpack-fluid-loader/src/loader.ts (hunk 6 above)
Comment (sdeshpande3) [reply packages/tools/webpack-fluid-loader/src/loader.ts:96]: nvm...it should be synchronous

File: packages/loader/container-loader/src/test/containerContext.spec.ts (hunk 10 above)
Comment (sdeshpande3) [reply packages/loader/container-loader/src/test/containerContext.spec.ts:190]: IFluidCodeDetailsComparer has two required functions: satisfies and compare. After typing ICodeDetailsLoader, compare was missing.

Comment (ssimic2) [reply packages/loader/container-loader/src/test/containerContext.spec.ts:190]: Not sure I fully understand, but it sounds like you are clear on the purpose and validity of this test.

**Review by ssimic2 — APPROVED** [review]

**Comment by tylerbutler** [comment]
Code scanning issues are not related to this change, so I'm overriding the failure and merging.

**Comment by sdeshpande3** [comment]
Thank you @tylerbutler!
